Mathlib/NumberTheory/NumberField/Units.lean

```diff
@@ -149,4 +161,398 @@ theorem rootsOfUnity_eq_torsion [NumberField K] :
+  refine le_antisymm (le_top) ?_
+  -- The standard basis
+  let B := Pi.basisFun ℝ {w : InfinitePlace K // w ≠ w₀}
+  -- The image by log_embedding of the family of units constructed above
+  let v := fun w : { w : InfinitePlace K // w ≠ w₀ } => logEmbedding K (exists_unit K w).choose
+  -- To prove the result, it is enough to prove that the family `v` is linearly independent
+  suffices B.det v ≠ 0 by
+    rw [← isUnit_iff_ne_zero, ← is_basis_iff_det] at this
+    rw [← this.2]
+    exact Submodule.span_monotone (fun _ ⟨w, hw⟩ =>
+      ⟨(exists_unit K w).choose, trivial, by rw [← hw]⟩)
+  rw [Basis.det_apply]
+  -- We use a specific lemma to prove that this determinant is nonzero
+  refine det_ne_zero_of_sum_col_lt_diag (fun w => ?_)
+  simp_rw [Real.norm_eq_abs, Basis.coePiBasisFun.toMatrix_eq_transpose, Matrix.transpose_apply]
+  rw [← sub_pos, Finset.sum_congr rfl (fun x hx => abs_of_neg ?_), Finset.sum_neg_distrib,
+    sub_neg_eq_add, Finset.sum_erase_eq_sub (Finset.mem_univ _), ← add_comm_sub]
+  refine add_pos_of_nonneg_of_pos ?_ ?_
+  · rw [sub_nonneg]
+    exact le_abs_self _
+  · rw [logEmbedding_sum_component (exists_unit K w).choose]
+    refine mul_pos_of_neg_of_neg ?_ ((exists_unit K w).choose_spec _ w.prop.symm)
+    rw [mult]; split_ifs <;> norm_num
+  · refine mul_neg_of_pos_of_neg ?_ ((exists_unit K w).choose_spec x ?_)
+    · rw [mult]; split_ifs <;> norm_num
+    · exact Subtype.ext_iff_val.not.mp (Finset.ne_of_mem_erase hx)
+
+end span_top
+
+/-- The unit rank of the number field `K`, it is equal to `card (InfinitePlace K) - 1`. -/
+def _root_.NumberField.Units.rank : ℕ := Fintype.card (InfinitePlace K) - 1
+
+open FiniteDimensional
+
+theorem unitLattice_discrete : DiscreteTopology (unitLattice K) := by
+  refine discreteTopology_of_open_singleton_zero ?_
+  refine isOpen_singleton_of_finite_mem_nhds 0 (s := Metric.closedBall 0 1) ?_ ?_
+  · exact Metric.closedBall_mem_nhds _ (by norm_num)
+  · refine Set.Finite.of_finite_image ?_ (Set.injOn_of_injective Subtype.val_injective _)
+    convert unitLattice_inter_ball_finite K 1
+    ext x
+    refine ⟨?_, fun ⟨hx1, hx2⟩ => ⟨⟨x, hx1⟩, hx2, rfl⟩⟩
+    rintro ⟨x, hx, rfl⟩
+    exact ⟨Subtype.mem x, hx⟩
+
+protected theorem finrank :
+    finrank ℝ ({w : InfinitePlace K // w ≠ w₀} → ℝ) = Units.rank K := by
+  simp only [finrank_fintype_fun_eq_card, Fintype.card_subtype_compl,
+    Fintype.card_ofSubsingleton, rank]
+
+theorem unitLattice_moduleFree : Module.Free ℤ (unitLattice K) := by
+  have := unitLattice_discrete K
+  exact Zlattice.module_free ℝ (unitLattice_span_eq_top K)
+
+theorem unitLattice_moduleFinite : Module.Finite ℤ (unitLattice K) := by
+  have := unitLattice_discrete K
+  exact Zlattice.module_finite ℝ (unitLattice_span_eq_top K)
+
+theorem unitLattice_rank : finrank ℤ (unitLattice K) = Units.rank K := by
+  have := unitLattice_discrete K
+  rw [← dirichlet.finrank]
+  exact Zlattice.rank ℝ (unitLattice_span_eq_top K)
+
+end dirichlet
```

Comment:
The three results above should be outside the namespace, I think. They are main outcomes of all the work above, and they are also results that will be used in applications of this file, just like the results below.